src/v/raft/tests/persisted_stm_test.cc

```diff
@@ -766,3 +774,85 @@ TEST_F_CORO(state_machine_fixture, test_concurrent_apply_and_snapshot) {
     co_await std::move(write_sleep_f);
     co_await std::move(local_snapshot_f);
 }
+
+/**
+ * Tests the scenario in which an STM is added to the partition after it was
+ * already alive and Raft snapshot was taken on the partition.
+ *
+ * The snapshot doesn't contain data for the newly created stm, however the stm
+ * next offset should still be updated to make it possible for the STM to catch
+ * up.
+ */
+TEST_F_CORO(persisted_stm_test_fixture, test_snapshot_in_background_apply) {
+    ss::shared_ptr<other_persisted_kv> other_stm;
+    for (int i = 0; i < 3; ++i) {
+        add_node(model::node_id(i), model::revision_id(0));
+    }
+    for (auto& [_, node] : nodes()) {
+        co_await node->initialise(all_vnodes());
+        raft::state_machine_manager_builder builder;
+        auto stm = builder.create_stm<persisted_kv>(*node);
+        other_stm = builder.create_stm<other_persisted_kv>(*node);
+        co_await node->start(std::move(builder));
+        node_stms.emplace(node->get_vnode(), std::move(stm));
+    }
+
+    kv_state expected;
+    auto ops = random_operations(2000);
+    for (auto batch : ops) {
+        co_await apply_operations(expected, std::move(batch));
+    }
+
+    co_await wait_for_apply();
+    for (const auto& [_, stm] : node_stms) {
+        ASSERT_EQ_CORO(stm->state, expected);
+    }
+
+    // take local snapshot on every node
+    co_await take_local_snapshot_on_every_node();
+    // update state
+    auto ops_phase_two = random_operations(50);
+    for (auto batch : ops_phase_two) {
+        co_await apply_operations(expected, std::move(batch));
+    }
+
+    co_await wait_for_apply();
+    for (const auto& [_, stm] : node_stms) {
+        ASSERT_EQ_CORO(stm->state, expected);
+    }
+    // force background apply on one of the state machines
+    auto throwing_stm = node_stms.begin()->second;
+    throwing_stm->set_throw_on_apply(true);
+
+    auto ops_phase_three = random_operations(200);
+    for (auto batch : ops_phase_three) {
+        co_await apply_operations(expected, std::move(batch));
+    }
+
+    auto leader = co_await wait_for_leader(10s);
+    auto offset = node(leader).raft()->dirty_offset();
+    co_await wait_for_committed_offset(
+      node(leader).raft()->dirty_offset(), 10s);
+
+    // while one of the stms is running in the background we are going to take
+    // raft snapshot
+    std::optional<state_machine_manager::snapshot_result> snapshot;
+    // take snapshot on one of the nodes that state machine is fully up to date.
+    for (auto [id, stm] : node_stms) {
+        if (stm->last_applied() == offset) {
+            snapshot
+              = co_await node(id.id()).raft()->stm_manager()->take_snapshot();
```

Comment:
break here to conform to the comment that says we take a snapshot on _one_ node?